patches/store-socket.diff

```diff
@@ -24,14 +24,36 @@ Index: code-server/lib/vscode/src/vs/workbench/api/node/extHostExtensionService.
  import * as performance from 'vs/base/common/performance';
  import { createApiFactoryAndRegisterActors } from 'vs/workbench/api/common/extHost.api.impl';
  import { RequireInterceptor } from 'vs/workbench/api/common/extHostRequireInterceptor';
-@@ -72,6 +74,10 @@ export class ExtHostExtensionService ext
- 		if (this._initData.remote.isRemote && this._initData.remote.authority) {
- 			const cliServer = this._instaService.createInstance(CLIServer);
- 			process.env['VSCODE_IPC_HOOK_CLI'] = cliServer.ipcHandlePath;
-+
-+			fs.writeFile(path.join(os.tmpdir(), 'vscode-ipc'), cliServer.ipcHandlePath).catch((error) => {
-+				this._logService.error(error);
-+			});
- 		}
+@@ -17,6 +19,7 @@ import { ExtensionRuntime } from 'vs/wor
+ import { CLIServer } from 'vs/workbench/api/node/extHostCLIServer';
+ import { realpathSync } from 'vs/base/node/extpath';
+ import { ExtHostConsoleForwarder } from 'vs/workbench/api/node/extHostConsoleForwarder';
++import { IExtHostWorkspace } from '../common/extHostWorkspace';
+ 
+ class NodeModuleRequireInterceptor extends RequireInterceptor {
  
- 		// Module loading tricks
+@@ -79,6 +82,24 @@ export class ExtHostExtensionService ext
+ 		await interceptor.install();
+ 		performance.mark('code/extHost/didInitAPI');
+ 
++		(async () => {
++			let socketPath = process.env['VSCODE_IPC_HOOK_CLI'];
++			if (!socketPath) {
++				return;
++			}
++			const workspace = this._instaService.invokeFunction((accessor) => {
++				const workspaceService = accessor.get(IExtHostWorkspace);
++				return workspaceService.workspace;
++			});
++			const entry = {
++				workspace,
++				socketPath
++			};
++			fs.appendFile(path.join(os.tmpdir(), 'vscode-ipc'), '\n' + JSON.stringify(entry), 'utf-8');
```

Comment:
> The newline-delimited JSON does seem a bit odd to me. What sort of consistency issues did you see with reading the file first compared to only writing?

I didn't encounter consistency issues in practice, but without append, one can imagine opening two editors simultaneously (e.g. chrome restores last browsing session) and a write undoing the previous write (i.e. good: `r w r w`, bad: `r r w w`), leading to an orphan session.

I too would prefer JSON, but [Newline-delimited JSON](https://en.wikipedia.org/wiki/JSON_streaming#Newline-Delimited_JSON) isn't the weirdest thing :) There are node packages for parsing them, but depending on them might be overkill...


> If we are worried about multiple things reading/writing the file at the same time maybe we could have separate files per workspace.

This would still lead to potential consistency problems within a workspace (e.g. two tabs open to the same workspace), if we don't want orphan sessions (https://github.com/coder/code-server/pull/6191#issuecomment-1542592138)

> Also I think appending like this will cause us to endlessly accumulate socket paths? I know we have the deletion code but that only deletes until it finds a good socket and since the good socket will always be the latest we will never clean up the old ones. Plus that code only runs when you are trying to open a file via the CLI which some folks may never do. Changing how we write here would more or less remove the need to clean up there.

Addressing this in https://github.com/coder/code-server/pull/6191#issuecomment-1542722947

---

I personally think removing the possibility of races here is worth using line-delimited JSON over JSON, but I wouldn't be opposed to just using JSON and live with read-write races as a theoretical possibility.

Reply:
Good points!  I agree with line-delimited as the better choice over races but in the case of multiple tabs open to the same workspace maybe it does not matter too much which one wins.